File: Service/src/main/java/contestUtils/ContestServices.java
package contestUtils;

import example.example.*;
import example.model.Event;
import example.model.Office;
import example.model.Participant;
import example.model.User;
import example.repo.EntityRepoException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import DTO.EventDTO;
import DTO.ParticipantDTO;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ContestServices implements IContestServices{
    private final UserService userService;
    private final ParticipantService participantService;
    private final EventService eventService;
    private final Map<String, IMainObserver> loggedClientes;
    private static final int DEFAULT_THREADS = 8;
    private static final Logger logger = LogManager.getLogger();

    public ContestServices(EventService eventService, ParticipantService participantService, UserService userService) {
        this.eventService = eventService;
        this.participantService = participantService;
        this.userService = userService;
        loggedClientes = new ConcurrentHashMap<>();
    }

    @Override
    public synchronized User login(String userName, String password, IMainObserver client) throws Exception {
        Optional<User> foundUser = userService.getLogin(userName, password);
        if (foundUser.isPresent()) {
            if(loggedClientes.get(userName) != null) {
                logger.error("User already logged in");
                throw new Exception("User already logged in");
            }
            loggedClientes.put(userName, client);
        }
        else{
            logger.error("Authentication failed");
            throw new Exception("Authentication failed");
        }
        return foundUser.get();
    }

    @Override
    public synchronized void logout(User user, IMainObserver client) throws Exception {
        IMainObserver localClient = loggedClientes.remove(user.getUserName());
        if(localClient == null) {
            logger.error("User: " + user.getUserName() + " is not logged in");
            throw new Exception("User : " + user.getUserName() + " is not logged in");
        }
    }

    @Override
    public synchronized Collection<EventDTO> getEventsWithParticipantsCount() throws EntityRepoException {
        return eventService.getEventsWithParticipantsCount();
    }

    @Override
    public synchronized Collection<ParticipantDTO> getParticipantsForEventWithCount(Long eventId) throws EntityRepoException {
        return eventService.getParticipantsForEventWithCount(eventId);
    }

    @Override
    public void saveEventEntry(List<Office> offices) {
        for (Office office : offices) {
            try {
                eventService.saveEventEntry(office);
            }catch(EntityRepoException e) {
                logger.error("Failed to save event entry for office: " + office, e);
            }
        }

        ExecutorService executor = Executors.newFixedThreadPool(DEFAULT_THREADS);
        for (IMainObserver client : loggedClientes.values()) {
            executor.execute(() -> {
                try {
                    client.newRegistration(eventService.getEventsWithParticipantsCount().stream().toList());
                } catch (Exception e) {
                    logger.error(e);
                }
            });
        }
        executor.shutdown();
    }

    @Override
    public synchronized Collection<Participant> findAllParticipants() throws EntityRepoException {
        return participantService.getAll();
    }


    @Override
    public void saveParticipant(Participant participant) {
        try {
            participantService.add(participant);
            ExecutorService executor = Executors.newFixedThreadPool(DEFAULT_THREADS);
            for (IMainObserver client : loggedClientes.values()) {
                executor.execute(() -> {
                    try {
                        client.participantAdded(participant);
                    } catch (Exception e) {
                        logger.error(e);
                    }
                });
            }
            executor.shutdown();
        }catch(EntityRepoException e) {
            logger.error("Failed to save participant", e);
        }
    }

    @Override
    public synchronized Collection<Event> findAllEvents() throws Exception {
        try {
            return eventService.getAll();
        } catch (EntityRepoException e) {
            logger.error("Failed to load events", e);
            throw new Exception("Could not load events");
        }
    }

    @Override
    public void saveEvent(Event event) throws Exception {

    }
}

File: java-app/Service/ContestServices.cs
using mpp_proiect_csharp_DianaGliga11.Model;
using mpp_proiect_csharp_DianaGliga11.Model.DTO;
using log4net;

namespace Service;

public class ContestServices : IContestServices
{
    private readonly I_UserService _userService;
    private readonly I_ParticipantService _participantService;
    private readonly I_EventService _eventService;
    private readonly IDictionary<string, IMainObserver> _loggedClients;
    private static readonly ILog Log = LogManager.GetLogger(typeof(ContestServices));

    public ContestServices(I_UserService userService, I_ParticipantService participantService, I_EventService eventService)
    {
        this._eventService = eventService;
        this._participantService = participantService;
        this._userService = userService;
        _loggedClients = new Dictionary<string, IMainObserver>();
    }

    public User Login(string username, string password, IMainObserver client)
    {
        User? foundUser = _userService.getLogin(username, password);
        if (foundUser != null)
        {
            if (_loggedClients.ContainsKey(foundUser.UserName))
            {
                Log.Error("User already log in");
                throw new Exception("User already logged in.");
            }
            _loggedClients[foundUser.UserName] = client;
            return foundUser;
        }
        else
        {
            Log.Error("Authentication failed");
            throw new Exception("Authentication failed.");
        }
    }

    public void Logout(User user, IMainObserver client)
    {
        if (!_loggedClients.TryGetValue(user.UserName, out var localClient))
        {
            Log.Error("User is not logged in");
            throw new Exception("User is not logged in.");
        }

        _loggedClients.Remove(user.UserName);
    }

    public List<EventDTO> GetEventsWithParticipantsCount()
    {
        return _eventService.getEventsWithParticipantsCount().ToList();
    }

    public List<ParticipantDTO> GetParticipantsForEventWithCount(long eventId)
    {
        return _eventService.getParticipantsForEventWithCount(eventId).ToList();
    }

    public List<Participant> GetAllParticipants()
    {
        return _participantService.getAll().ToList();
    }

    public List<Event> GetAllEvents()
    {
        return _eventService.getAll().ToList();
    }

    public void SaveEventsEntries(List<Office> newEntries)
    {
        foreach (Office entry in newEntries)
        {
            _eventService.saveEventEntry(entry);
        }

        var updatedEvents= GetEventsWithParticipantsCount();
        foreach (IMainObserver client in _loggedClients.Values)
        {
           client.EventEvntriesAdded(updatedEvents);
        }
    }
    public void SaveParticipant(Participant participant, IMainObserver sender)
    {
        _participantService.add(participant);
        foreach (IMainObserver client in _loggedClients.Values)
        {
            if (client != sender)
            {
               client.ParticipantAdded(participant);
            }
        }
    }
}
